Bugzilla/User.pm

```diff
@@ -2877,6 +2877,43 @@ sub user_id_to_login {
   return $login || '';
 }
 
+sub validate_password {
+  my $check = validate_password_check(@_);
+  ThrowUserError($check) if $check;
+  return 1;
+}
+
+sub validate_password_check {
+  my ($password, $matchpassword) = @_;
+
+  if (length($password) < USER_PASSWORD_MIN_LENGTH) {
+    return 'password_too_short';
+  }
+  elsif ((defined $matchpassword) && ($password ne $matchpassword)) {
+    return 'password_mismatch';
+  }
+
+  my $complexity_level = Bugzilla->params->{password_complexity};
+  if ($complexity_level eq 'bmo') {
+    my $features = 0;
+
+    # A password can be 3 (or more) or more 4 (or more) letter words
+    if ($password =~ /\W{4}\s+\W{4}\s+\W{4}/) {
```

Comment:
This regex is wrong; it's matching on *non-word* characters (so `"""" """" """"` passes but `test test test` doesn't), isn't anchored to the start or end of the string, only matches words that are exactly four chars long, and Perl's definition of "word" (`[0-9a-zA-Z_]`) shouldn't be used as it's English only and doesn't include chars such as `'` that exist in words.

We probably want something like `/^\S{4,}(?:\s+\S{4,}){2,}/`

Please add tests for lots of variations of passwords.

Reply:
Thanks for catching that. I meant to use \w instead of \W. You regex definitely looks better so I will try that and also update the automated tests.